diff --git a/src/Networking.h b/src/Networking.h
index e26888c..1b35977 100644
--- a/src/Networking.h
+++ b/src/Networking.h
@@ -29,6 +29,10 @@
 #pragma comment(lib, "ws2_32.lib")
 #define SHUT_WR SD_SEND
 #ifdef __MINGW32__
+extern "C" {
+	WINSOCK_API_LINKAGE INT WSAAPI inet_pton(INT, PCSTR, PVOID);
+	WINSOCK_API_LINKAGE PCSTR WSAAPI inet_ntop(INT, PVOID, PSTR, size_t);
+}
 // Windows has always been tied to LE
 #define htobe64(x) __builtin_bswap64(x)
 #define be64toh(x) __builtin_bswap64(x)
